exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,382 @@
+{
+    Cursors = dictionary_createDictionary();
+}
+
+void sqlite_shutdownModule(void)
+{
+    dictionary_freeDictionary(Cursors);
+}
+
+
+static Object *check_database(TExecutor *exec, Object *pdb)
+{
+    if (pdb == NULL || pdb->ptr == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        return NULL;
+    }
+    return pdb;
+}
+
+
+void object_releaseDatabaseObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                assert(sqlite3_close(o->ptr) == SQLITE_OK);
+                o->ptr = NULL;
+            }
+            free(o);
+        }
+    }
+}
+
+Cell *object_databaseObjectToString(Object *self)
+{
+    char s[32];
+    snprintf(s, 32, "<SQLITE %" PRId64 ">", (int64_t)self->ptr);
+    Cell *r = cell_fromCString(s);
+    return r;
+}
+
+Object *object_createDatabaseObject(sqlite3 *db)
+{
+    Object *r = object_createObject();
+    r->ptr = db;
+    r->release = object_releaseDatabaseObject;
+    r->toString = object_databaseObjectToString;
+
+    return r;
+}
+
+
+
+typedef struct tagTCursor {
+    TExecutor *exec;
+    sqlite3 *db;
+    char *query;
+    sqlite3_stmt *stmt;
+    int columns;
+} Cursor;
+
+static void openCursor(Cursor *self)
+{
+    int r = sqlite3_prepare_v2(self->db, self->query, -1, &self->stmt, NULL);
+    if (r != SQLITE_OK) {
+        self->exec->rtl_raise(self->exec , "SqlException", sqlite3_errmsg(self->db), number_from_sint32(r));
+        return;
+    }
+    self->columns = sqlite3_column_count(self->stmt);
+}
+
+static BOOL fetchCursor(Cursor *self, Cell *result)
+{
+    int r = sqlite3_step(self->stmt);
+    if (r == SQLITE_DONE) {
+        return FALSE;
+    }
+    if (r == SQLITE_ROW) {
+        for (int i = 0; i < self->columns; i++) {
+            Cell *col = cell_fromCString((const char*)sqlite3_column_text(self->stmt, i));
+            cell_arrayAppendElementPointer(result, col);
+        }
+    } else {
+        self->exec->rtl_raise(self->exec, "SqlException", sqlite3_errmsg(self->db), number_from_sint32(r));
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static void closeCursor(Cursor *self)
+{
+    sqlite3_finalize(self->stmt);
+    self->stmt = NULL;
+}
+
+static void freeCursor(Cursor *self)
+{
+    if (self->stmt) {
+        closeCursor(self);
+    }
+    free(self->query);
+    free(self);
+}
+
+static Cursor *declareCursor(TExecutor *exec, sqlite3 *db, char *query)
+{
+    Cursor *r = malloc(sizeof(Cursor));
+    if (r == NULL) {
+        fatal_error("Could not alloc SQLite Cursor.\n");
+    }
+
+    r->db = db;
+    r->query = query;
+    r->stmt = NULL;
+    r->columns = 0;
+    r->exec = exec;
+
+    return r;
+}
+
+static int callback(void *rowscell, int columns, char **values, char **names)
+{
+    (void)names; // prevent compiler warning "expected formal parameter list, not a type list"
```

Comment:
This comment doesn't make sense. Isn't this statement here to indicate an unused parameter?

Reply:
Corrected.